### DISCUSSION TITLE
How to memoize rows correctly with `@tanstack/react-table`

### DISCUSSION THREAD

**Discussion by jonahallibone** [TanStack/virtual#535] [Q&A] [answered]
My code below is pretty representative of how I am rendering my table in my app. My problem comes when I wrap `memo(GeneralTableRows)`, I suddenly lose access to all state updates in the table. Things like `getIsSelected()` etc, just cease to function. I assume it's because of the memoized nature of `useReactTable` having a stable reference. This makes sense, sort of, but it causes all sorts of problem relating to optimizing render updates with `@tanstack/react-virtual`. Without memoizing each row, every single row updates on every scroll, which causes terrible rendering performance as my rows have somewhat complex components in the cell. The desired result is that the already rendered rows do not rerender on each scroll, and only rerender when:

a) interacted with 
b) rendered into the list 
c) unmounted from the screen. 

Any help on this would be great. 

As an aside, I have managed to get it to the point where the rows but not the cells rerender, however, this still seems suboptimal and also feels like I must be doing something wrong. My instinct is that my keys are wrong inside mapping of the rows, but I can't be sure.
```
const GeneralTableRows = <T>({
  rows,
}: GeneralTableRowsProps<T>) => {
  const rowCount = rows.length;

  const rowVirtualizer = useVirtualizer({
    count: rowCount,
    getScrollElement: () => tableContainerRef?.current ?? null,
    getItemKey: (index) => rows[index]?.id ?? index,
    estimateSize: () => 37,
  });

  const width = ${
    table.getTotalSize() - table.getVisibleLeafColumns().length
  }px`;

  const totalSize = rowVirtualizer.getTotalSize();

  const virtualRows = rowVirtualizer.getVirtualItems();

  return virtualRows.map((virtualRow) => {
    return (
     <GeneralTableRows
       rows={rows}
       key={virtualRow.key}
       virtualRow={virtualRow}
       measureElement={rowVirtualizer.measureElement}
       start={virtualRow.start}
     />
    );
  });
};
```

**Comment by piecyk** [comment]
First `getItemKey` should be stale, it's part of re-create the measurements array. If you could create small codesandbox then i can have a look

> As an aside, I have managed to get it to the point where the rows but not the cells rerender, however, this still seems suboptimal and also feels like I must be doing something wrong. 

Overall in table most cases you want to memozie rows.

**Reply by jonahallibone** [reply]
I was working on trying to get you a repro but then codesandbox freaked out, seems i need to remake it. While I rebuild that example, I'd love some clarity on the "`getItemKey` should be stale" comment? The docs don't state whether we should use the `virtualItem.key` at all, but I assumed we'd want to? It doesn't explicitly state if it's for internal or external usage. The examples I saw use `virtualItem.index` when keying in render, but in the `@tanstack/react-table` examples they use `row.id` for keying in render. Overall, this leaves me unsure how best to virtualize and memoize rows when relating to `key` as a prop. It also leaves me unsure of what the recommended approach is. Assuming the majority of people will be wanting to use `@tanstack/virtual` with `@tanstack/table` it would be great to have clarity on these things. An example in either repo would be great.

Also, I agree about memoizing rows. I want to memoize rows, but I think my keys are either incorrect or the stable reference being emitted by the table doesn't let me memoize while still being reactive to state changes.

**Reply by jonahallibone** [reply]
Here's an example: https://codesandbox.io/p/sandbox/xenodochial-wescoff-lddjs1?file=%2Fcomponents%2Fgeneral-table-rows.tsx&selection=%5B%7B%22endColumn%22%3A59%2C%22endLineNumber%22%3A140%2C%22startColumn%22%3A59%2C%22startLineNumber%22%3A140%7D%5D

**Reply by jonahallibone** [reply]
https://lddjs1-3000.csb.app/

This is a preview you can run the Component Inspector on.

**Reply by jonahallibone** [reply]
@piecyk No rush but just curious if you've had a second to look at this!

**Reply by obinmatt** [reply]
@piecyk I am also interested in your feedback here, when you get the chance :)

**Reply by piecyk** [reply]
@jonahallibone yes we need to update docs, `getItemKey` needs to be stale, so by this i meant wrapping it with react `useCallback` as this should not change on every re-render.

**Reply by jonahallibone** [reply]
Do you have a quick snippet you could add as an example? Just to clarify your meaning?

**Reply by obinmatt** [reply]
@piecyk thank you!! adding the callback made a huge difference

**Reply by obinmatt** [reply, marked as answer]
@jonahallibone you want your getItemKey to be `React.useCallback((index: number) => rows[index]?.id ?? index, [rows])`
